intro/README.md

```diff
@@ -0,0 +1,561 @@
+     
+        int* p;
+        p = "hello" in aa;
+     
+        if (p)
+        {
+            *p = 4;  // update value associated with key; aa["hello"] == 4
+        }
+    }
+    ```
+    
+    The “in” expression yields a pointer to the value if the key is in the associative array, or null if not.
+
+    For more advanced operations on AAs check this [link](https://dlang.org/spec/hash-map.html#construction_assignment_entries).
+    For an exhaustive list of the AA properties check this [link](https://dlang.org/spec/hash-map.html#properties).
+
+
+Find the [majority element](https://leetcode.com/problems/majority-element/) in a string array using builtin associative arrays.
+You can start your implementation using the skeleton inside **majority-element** directory
+
+## Structs
+
+In D, structs are simple aggregations of data and their associated operations on that data:
+
+```d
+struct Rectangle
+{
+    size_t length, width;
+    int id;
+ 
+    size_t area() { return length*width; }
+    size_t perimeter() { return 2*(length + width); }
+    size_t isSquare() { return length == width; }
+    void setId(int id) { this.id = id; }
+}
+```    
+
+## Classes
+
+In D, classes are very similar with java classes:
+
+1. Classes can implement any number of interfaces
+2. Classes can inherit a single class
+3. Class objects are instantiated by reference only
+4. super has the same meaning as in Java
+5. overriding rules are very similar
+
+The fundamental difference between **structs** and **classes** is that the former are **value** types, while the latter are **reference** types.
+This means that whenever a struct is passed as an argument to an lvalue function parameter, the function will operate on a copy of the object.
+When a class is passed as an argument to an lvalue function parameter, the function will receive a reference to the object.
+
+Both **structs** and **classes** will be covered more in depth in a further lab.
+
+Navigate to the **demo/struct-class** directory.
+You will find 2 files implementing the same program.
+One uses a struct as a data container, while the other uses a class.
+
+- Compile both files and measure the run time of each program. How do you explain the differences?
+- How do you explain the output differences? How we can change the code for structs so both versions produce the same output?
+
+[Quiz](./quiz/struct-class.md)
+
+## Functions
+
+Functions are declared the same as in C. In addition, D offers some convenience features like:
+
+1. Uniform function call syntax (UFCS): allows that any call to a free function **fun(a)** can be written as member function call **a.fun()**
+
+    ```d
+    import std.algorithm : group;
+    import std.range : chain, dropOne, front, retro;
+    [1, 2].chain([3, 4]).retro; // 4, 3, 2, 1
+    [1, 1, 2, 2, 2].group.dropOne.front; // (2, 3)
+     
+    front(dropOne(group([1, 1, 2, 2, 2])));
+    ```
+
+D has a standard library called [phobos](https://dlang.org/phobos/).
```

Comment:
This paragraph is very abrupt. You could maybe put the exercise first and then instruct them to search for functions  in the standard library. Something along the lines of: "Find the median [...]. Use any number of functions from the [standard library](link). Use UFCS to make the code more expressive."